File: scripts/walfix.mjs
#!/usr/bin/env node
// Rescue tool: convert a SQLite database from WAL journaling back to the
// default DELETE journaling. Use when the Electron app crashes at startup
// with `SQLite3Error: unable to open database file` because some other
// tool (rusqlite, sqlite3 CLI, better-sqlite3, etc.) has flipped the
// file's header bytes 18/19 to 2/2 (the WAL marker).
//
// node-sqlite3-wasm's custom VFS has sqlite3_io_methods.iVersion=1, no
// shared-memory hooks — so it cannot open WAL-tagged files. See
// .claude/skills/sqlite-wal/ for the full constraint.
//
// Uses the built-in node:sqlite module (Node 22.5+, stable in Node 24+),
// which links real native SQLite and supports WAL.
//
// Usage: node scripts/walfix.mjs <path-to.db>
//
// Back up the DB first if you care; this script doesn't.

import { DatabaseSync } from 'node:sqlite';
import { existsSync, openSync, readSync, closeSync } from 'node:fs';

const dbPath = process.argv[2];
if (!dbPath) {
  console.error('usage: node scripts/walfix.mjs <path-to.db>');
  process.exit(2);
}
if (!existsSync(dbPath)) {
  console.error(`not found: ${dbPath}`);
  process.exit(2);
}

function readJournalHeader(p) {
  const fd = openSync(p, 'r');
  const buf = Buffer.alloc(20);
  readSync(fd, buf, 0, 20, 0);
  closeSync(fd);
  return { write_v: buf.readUInt8(18), read_v: buf.readUInt8(19) };
}

const before = readJournalHeader(dbPath);
console.log(`before: write_v=${before.write_v} read_v=${before.read_v}`);
if (before.write_v === 1 && before.read_v === 1) {
  console.log('already in DELETE mode — nothing to do');
  process.exit(0);
}

let db;
try {
  db = new DatabaseSync(dbPath);
} catch (e) {
  console.error('open failed:', e.message);
  process.exit(1);
}

// Checkpoint any pending WAL frames into the main DB before downgrading
// the journal mode. TRUNCATE writes them and clears the -wal sidecar.
try {
  db.exec('PRAGMA wal_checkpoint(TRUNCATE)');
} catch (e) {
  console.warn('checkpoint warning:', e.message);
}

const after = db.prepare('PRAGMA journal_mode = DELETE').get();
db.close();

const headerAfter = readJournalHeader(dbPath);
console.log(`after:  write_v=${headerAfter.write_v} read_v=${headerAfter.read_v}`);
console.log(`journal_mode now: ${after.journal_mode}`);

if (after.journal_mode !== 'delete' || headerAfter.write_v !== 1 || headerAfter.read_v !== 1) {
  console.error('downgrade did not complete cleanly');
  process.exit(1);
}

File: tauri-spike/src-tauri/src/bin/walfix.rs
// One-shot rescue tool: convert a SQLite database from WAL journaling
// back to the default DELETE journaling. Use only when something
// (such as the Tauri spike) has WAL-tagged a file that node-sqlite3-wasm
// then refuses to open with SQLITE_CANTOPEN — see the investigation in
// docs/plans/tauri-port-evaluation-baseline.md.
//
// Usage: walfix <path-to.db>
//
// The operation is destructive only of the journal-mode header bytes
// (write_version=2 → 1, read_version=2 → 1). All authored rows are
// preserved. Take a backup first if you care; this binary does not.

use rusqlite::Connection;
use std::env;
use std::process::ExitCode;

fn main() -> ExitCode {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("usage: walfix <path-to.db>");
        return ExitCode::from(2);
    }
    let path = &args[1];
    let conn = match Connection::open(path) {
        Ok(c) => c,
        Err(e) => {
            eprintln!("open {path}: {e}");
            return ExitCode::from(1);
        }
    };
    // Checkpoint any pending WAL frames into the main DB first; without this,
    // setting journal_mode=DELETE on a hot WAL would lose anything not yet
    // checkpointed. PASSIVE checkpoint is fine because no other process
    // should hold the file (caller's responsibility).
    if let Err(e) = conn.pragma_update(None, "wal_checkpoint", "TRUNCATE") {
        eprintln!("checkpoint warning: {e}");
    }
    let mode: String = match conn.pragma_update_and_check(
        None, "journal_mode", "DELETE", |row| row.get(0),
    ) {
        Ok(s) => s,
        Err(e) => {
            eprintln!("set journal_mode=DELETE: {e}");
            return ExitCode::from(1);
        }
    };
    println!("journal_mode now: {mode}");
    if mode.to_lowercase() != "delete" {
        eprintln!("expected 'delete' but got '{mode}'");
        return ExitCode::from(1);
    }
    ExitCode::SUCCESS
}
